src/main/java/com/snowflake/kafka/connector/internal/streaming/schemaevolution/iceberg/IcebergSchemaEvolutionService.java

```diff
@@ -17,48 +20,163 @@ public class IcebergSchemaEvolutionService implements SchemaEvolutionService {
     this.conn = conn;
-    this.tableSchemaResolver = new IcebergTableSchemaResolver();
-  }
-
-  @VisibleForTesting
-  IcebergSchemaEvolutionService(
-      SnowflakeConnectionService conn, TableSchemaResolver tableSchemaResolver) {
-    this.conn = conn;
-    this.tableSchemaResolver = tableSchemaResolver;
+    this.icebergTableSchemaResolver = new IcebergTableSchemaResolver();
+    this.mergeTreeService = new IcebergColumnTreeMerger();
+    this.typeBuilder = new IcebergColumnTreeTypeBuilder();
   }
 
   /**
-   * Execute an ALTER TABLE command if there is any extra column that needs to be added, or any
-   * column nullability that needs to be updated, used by schema evolution
-   *
-   * @param targetItems target items for schema evolution such as table name, columns to drop,
-   *     columns to add
-   * @param record the sink record that contains the schema and actual data
+   * @param targetItems column and field names from InsertError returned by ingest-sdk
+   * @param record record that caused an error
+   * @param existingSchema schema stored in a channel
    */
   @Override
-  public void evolveSchemaIfNeeded(SchemaEvolutionTargetItems targetItems, SinkRecord record) {
+  public void evolveSchemaIfNeeded(
+      SchemaEvolutionTargetItems targetItems,
+      SinkRecord record,
+      Map<String, ColumnProperties> existingSchema) {
     String tableName = targetItems.getTableName();
-    List<String> columnsToAdd = targetItems.getColumnsToAdd();
+
+    Set<String> columnsToEvolve = extractColumnNames(targetItems);
+
     // Add columns if needed, ignore any exceptions since other task might be succeeded
-    if (!columnsToAdd.isEmpty()) {
-      LOGGER.debug("Adding columns to iceberg table: {} columns: {}", tableName, columnsToAdd);
-      TableSchema tableSchema =
-          tableSchemaResolver.resolveTableSchemaFromRecord(record, columnsToAdd);
-      try {
-        conn.appendColumnsToIcebergTable(tableName, tableSchema.getColumnInfos());
-      } catch (SnowflakeKafkaConnectorException e) {
-        LOGGER.warn(
-            String.format(
-                "Failure altering iceberg table to add column: %s, this could happen when multiple"
-                    + " partitions try to alter the table at the same time and the warning could be"
-                    + " ignored",
-                tableName),
-            e);
-      }
+    if (!columnsToEvolve.isEmpty()) {
+      LOGGER.debug("Adding columns to iceberg table: {} columns: {}", tableName, columnsToEvolve);
+      // some of the column might already exist, and we will modify them, not create
+      List<IcebergColumnTree> alreadyExistingColumns =
+          icebergTableSchemaResolver.resolveIcebergSchemaFromChannel(
+              existingSchema, columnsToEvolve);
+
+      List<IcebergColumnTree> modifiedOrAddedColumns =
+          icebergTableSchemaResolver.resolveIcebergSchemaFromRecord(record, columnsToEvolve);
+
+      List<IcebergColumnTree> columnsToAdd =
```

Comment:
There is no need to iterate twice over the list. The columns from `modifiedOrAddedColumns` are either modified or added, right?

Reply:
Righ, however I found my approach simpler and didn't care very much about performance of that.

Create 2 lists and write an if to put an element into one or another list.

```
  private Pair<List<IcebergColumnTree>, List<IcebergColumnTree>> distinguish(
          List<IcebergColumnTree> alreadyExistingColumns,
          List<IcebergColumnTree> modifiedOrAddedColumns) {

    ArrayList<IcebergColumnTree> columnsToModify = new ArrayList<>();
    ArrayList<IcebergColumnTree> columnsToAdd = new ArrayList<>();
    
    for (IcebergColumnTree tree : modifiedOrAddedColumns) {
      if (alreadyExistingColumns.stream()
              .anyMatch(alreadyExisting -> alreadyExisting.getColumnName()
                      .equalsIgnoreCase(tree.getColumnName()))) {
        
        columnsToModify.add(tree);
      } else {
        columnsToAdd.add(tree);
      }
    }
    return Pair<List<IcebergColumnTree>, List<IcebergColumnTree>>(columnsToAdd, columnsToModify);
  }
```
But there is not "Pair" in JDK8. 
This approach is a bit more messy. Slightly more performant.


~(It's messy ->) The second alternative I see is to distinguish columnsToAdd, and then having columnsToAdd iterate over columns to evolve and again match them with modifiedOrAddedColumns list.~

Reply:
None is perfect.

Reply:
For me it's a matter of avoiding code duplication rather than optimizing the code execution time. I don't see any problem with creating a wrapper class for two lists.
Anyway it's not a blocker to me.